### PR TITLE
import mutable globals used in Asyncify pass

### PR DESCRIPTION
**Pull request by nokotan**
This PR will be a solution to #15594

emscripten Asyncify won't work properly in side modules, because the globals, __asyncify_state and __asyncify_data, are not synchronized between main-module and side-modules.

With <https://github.com/WebAssembly/binaryen/pull/4427>, `__asyncify_state` and `__asyncify_data` will be imported global in the instrumented wasm.

### CODE REVIEW

**Review by kripken — COMMENTED** [review]
Code lgtm in general. We'll need to wait for the binaryen PR to be reviewed, and then land and roll in, before we can use it here. And then we should add a test.

**Comment by nokotan** [comment]
Test case test_asyncify_side_module, added in this PR, won't pass when SIDE_MODULE and MAIN_MODULE are flipped, and I can't find why..

**Comment by nokotan** [comment]
Seems that functions in side modules cannot be a origin of asyncify rewinding point.
To overcome this, we should make exports in side modules instrumented by `Asyncify.instrumentWasmExports`.

**Comment by kripken** [comment]
@nokotan Interesting, yes, that makes sense - sounds like the flipped test found a real issue here.

Yes, I think that solution should work. Let us know if you run into problems.

**Comment by chkuendig** [comment]
As the reporter of #15594: Awesome news!  I tested this and noticed 3 things (to be taken with a grain of salt - I'm not very familiar  emscripten):

a) I noticed this didn't work for dlopen libs (as dlsym also adds the functions) - I tried to work around this with [chkuendig/emscripten/commit/81224917233dc38d7542f5f2442ec4b647d27a5f](https://github.com/chkuendig/emscripten/commit/81224917233dc38d7542f5f2442ec4b647d27a5f). But that might be somewhat hacky (I lack the in-depth knowledge to be certain about the side effects)

b) `-s ASYNCIFY_IMPORTS=["_Z8my_sleepi"]` doesn't seem to be required? For me it also worked if any side module  methods were only instrumented as part of the exports

c) There's still some issue if the side module calls back into the main module (e.g. there's a my_sleep_v2()   which is compiled into the main_module but called from the side module). I'm currently working on building a minimal sample to reproduce this.

**Comment by nokotan** [comment]
@[chkuendig](https://github.com/chkuendig) Now included your patch into this PR, thank you for your investigation!

**Comment by chkuendig** [comment]
I looked further into this updated the previous comment - I couldn't yet find any other issues. Might be worth though to add the code in #15594 as a test as well (for the dlopen use case).

**Comment by nokotan** [comment]
Added new test case `test_asyncify_dlfcn`.

Before applying the [patch], the test case fails with the error below.

```
Aborted(Assertion failed: Missing signature argument to addFunction: function() {
    Asyncify.exportCallStack.push(x);
    try {
      return original.apply(null, arguments);
    } finally {
      if (!ABORT) {
        var y = Asyncify.exportCallStack.pop();
        assert(y === x);
        Asyncify.maybeStopUnwind();
      }
    }
  })
```

This function is the function wrapper in [instrumentWasmExports]. With [Type Reflection Proposal for WebAssembly JS API], the `sig` property (describes function signiture; required in addFunction) can be added into this wrapper function.

Anyway, the [patch] makes sense to access to **raw** exported function, which is not instrumented by Asyncify, in the side modules.

Maybe I should consider the difference of `Module['_main']` and `Module['asm']['main']`. (Instrumented  functions to use in JavaScript side or not instrumented ones to use inside WebAssembly world.)

[patch]: https://github.com/chkuendig/emscripten/commit/81224917233dc38d7542f5f2442ec4b647d27a5f
[instrumentWasmExports]: https://github.com/emscripten-core/emscripten/blob/main/src/library_async.js#L109
[Type Reflection Proposal for WebAssembly JS API]: https://github.com/WebAssembly/js-types/blob/main/proposals/js-types/Overview.md#addition-of-webassemblyfunction

**Comment by nokotan** [comment]
LLVM optimization strips memory imports from wasm binary generated with simple test case, the new test cases with "O2", "O3" compilation flags is still failing.
Binaryen's asyncify pass expects that instrumented modules have memory imports and adds memory section if none.

**Comment by nokotan** [comment]
@kripken ready to be reviewed, please check the changes.

File: src/library_async.js
Comment (kripken) [inline src/library_async.js:208]: Please add a comment here, as it is not obvious why this is needed (I actually don't know right now).

File: src/library_dylink.js
Comment (kripken) [inline src/library_dylink.js:971]:
```suggestion
        return GOT[symbol].value;
```

File: tests/test_core.py
Comment (kripken) [inline tests/test_core.py:7911]: The file could be C which would avoid name mangling here (pass `force_c=True` to `dylink_test`)

File: emcc.py
Comment (kripken) [inline emcc.py:558]: I had not realized that this needs to be run on the main and the side modules, but I guess that makes sense... if so, should we call the param on the binaryen side `asyncify-relocatable`? Otherwise it seems odd to pass it for a main module here.

Comment (nokotan) [reply emcc.py:558]: Originally I've designed sharing globals with asyncify as that these globals are exported from a main module and are imported into side modules. When creating this patch in Emscripten side, I've found that it will be easier that making those globals are imported into both main and side modules, and that this implementation will be more consistent with `__stack_pointer`, `__memory_base` and `__table_base`, which are all imported into main/side modules with the option `-sRELOCATABLE`.

Comment (nokotan) [reply emcc.py:558]: So, I agree with renaming `asyncify-side-module` to `asyncify-relocatable` on the binaryen side.

File: src/library_async.js
Comment (sbc100) [reply src/library_async.js:208]: They also get added to asmLibraryArg which acts as a kind of global namespace

File: emcc.py
Comment (nokotan) [reply emcc.py:558]: I've created new pull request of this renaming option. <https://github.com/WebAssembly/binaryen/pull/4596>

File: src/library_dylink.js
Comment (kripken) [reply src/library_dylink.js:969]: Hmm, sorry, I still don't understand this part. This is general linking code, so I'm surprised to see changes here - unless asyncify is using something we've never needed before, and I'm not sure what that is.

Also, the linked comment talks about functions, and below we call `addFunction`, but this checks `!= 0` on `.value`, which confuses me? Is it checking if it is not a function or something like that? If so then `typeof GOT[symbol]` might be better?

Comment (chkuendig) [reply src/library_dylink.js:969]: I added this. Without it, addFunction will try to add the function which will at `setWasmTableEntry` because it's a JS function  and then triggers the assert a few lines later because it doesn't have a signature:

```
Aborted(Assertion failed: Missing signature argument to addFunction: function() {
                Asyncify.exportCallStack.push(x);
                try {
                  return original.apply(null, arguments);
                } finally {
                  if (!ABORT) {
                    var y = Asyncify.exportCallStack.pop();
                    assert(y === x);
                    Asyncify.maybeStopUnwind();
                  }
                }
```

When I retested this just now, I did however notice that just checking whether the symbol is in `GOT` (as proposed by me earlier) doesn't work if we have multiple libraries with the same symbols (similar issue to #16581 )

Comment (kripken) [reply src/library_dylink.js:969]: Ah, then maybe we need to add the signature for the function? As long as the function has one, the existing asyncify code should copy it even into the wrapper function (the one that does `original.apply` that you pasted here),

https://github.com/emscripten-core/emscripten/blob/651e76db65712d32c61cbcd8cd1488e22ad2a5f5/src/library_async.js#L97-L102

Adding the signature would look like this:
```diff
diff --git a/src/library.js b/src/library.js
index b1947fc69..8e4c1f66e 100644
--- a/src/library.js
+++ b/src/library.js
@@ -3294,6 +3294,7 @@ LibraryManager.library = {
 
   $setWasmTableEntry__internal: true,
   $setWasmTableEntry__deps: ['$wasmTableMirror'],
+  $setWasmTableEntry__sig: 'ir',
   $setWasmTableEntry: function(idx, func) {
     wasmTable.set(idx, func);
     wasmTableMirror[idx] = func;
```
However the second param is a reference. I wrote "r" for that, in the diff. @sbc100 do we already have a letter for reference types?

Taking a step back, though, why does that function end up in a call to `addFunction()`? C/C++ code wouldn't be calling it, so I'm confused.

Comment (sbc100) [reply src/library_dylink.js:969]: `$setWasmTableEntry` is a JS-only function, it cannot be called from native code or added to the wasm table.. so I don't think it can/should have a signature at all.

File: tests/test_core.py
Comment (chkuendig) [reply tests/test_core.py]: Sorry, I think a typo swallowed parts of my previous comments. The function being added isn't `setWasmTableEntry` - `setWasmTableEntry` is where the adding fails because no signature can be found. I clarified the above comment with bold/italic changes.

Attached the debugger at this point in question. You'll see that the function doesn't have a signature and is already added to GOT.
image

File: src/library_dylink.js
Comment (chkuendig) [reply src/library_dylink.js:969]: Sorry, I think a typo swallowed parts of my previous comments. The function being added isn't `setWasmTableEntry` - `setWasmTableEntry` is where the adding fails because no signature can be found. I clarified the above comment with bold/italic changes.

I attached the debugger at this point in question. You'll see that the function doesn't have a signature and is already added to GOT. (edit: I quickly checked -  it's been instrumented by `instrumentWasmExports` - this is line 17172 you see as `FunctionLocation`)
image

Comment (kripken) [reply src/library_dylink.js:969]: > This approach though fails if multiple libraries have the same symbols.

1. Do you mean multiple libraries implement the same symbols? That would be incorrect, I think, as we need to declare this asyncify code just once (in the main module, and import it from side modules).
2. Or do you mean multiple libraries define the symbol (and all import it from one place)? That's the common case in general, which should be heavily tested.

Because of these questions I'm not sure why this PR is hitting a new type of problem in dynamic linking, so I'm still missing something...

> I was experimenting with keeping a reference to the original/wasm as part of the instrumented method. See https://github.com/chkuendig/emscripten/commit/2ea11b4734de5656aaf3285145831e2bb9ad53ae. But that also seems a bit hackish.

Interesting. I think that might make sense, actually - doesn't seem like a bad hack to me. But I still don't see why it is needed. So, we do copy signatures to the asyncify wrappers, here:

https://github.com/emscripten-core/emscripten/blob/3a3f3579d09c01ceb9616c90339b9064206dc747/src/library_async.js#L97-L102

It seems like you need the opposite, for exports as opposed to imports... do you know why that is, what is new in this PR that needs that?

Comment (nokotan) [reply src/library_dylink.js:969]: The changes, <https://github.com/chkuendig/emscripten/commit/2ea11b4734de5656aaf3285145831e2bb9ad53ae>, look good for me. It seems that `orig` can be passed to `addFunction` and `dlsym_js` expects that function index are returned.

```diff
#if ASYNCIFY
      if ('orig' in result) {
-       result = result.orig;
+       result = addFunction(result.orig);
      }
#endif
```

Comment (nokotan) [reply src/library_dylink.js:969]: https://github.com/emscripten-core/emscripten/blob/3a3f3579d09c01ceb9616c90339b9064206dc747/src/library_async.js#L97-L102

This is only enabled with the flags `ASSERTION`, `ASYNCIFY` and `MAIN_MODULE`, and can be applied to JavaScript function imports of main module. I think that this is no effect to imported functions from side modules, because they don't have .sig property.

Comment (kripken) [reply src/library_dylink.js:969]: Ok, thanks, I re-read this now and I think I understand. Asyncify adds an extra JS wrapper around exports, and dynamic linking needs to see through that wrapper, basically. sgtm

File: src/library_async.js
Comment (kripken) [reply src/library_async.js:211]: @sbc100 So should we read from `asmLibraryArg` here, perhaps?

File: src/library_dylink.js
Comment (kripken) [inline src/library_dylink.js:970]:
```suggestion
      // Asyncify wraps exports, and we need to look through those wrappers.
      if ('orig' in result) {
```

File: tests/test_core.py
Comment (kripken) [reply tests/test_core.py:8069]: Interesting... If I understand, the issue is that

1. Normally `ASYNCIFY_IMPORTS` just contains JS functions, because those are the imports.
2. With dynamic linking, **a wasm function in one module can be an import in another**, and so `ASYNCIFY_IMPORTS` must contain wasm functions too, when they are linked across modules and when they end up on the stack when pausing.

Is that right? If so, we should document it as it is not obvious. Adding a note to `./site/source/docs/porting/asyncify.rst` would be good.

**Review by kripken — COMMENTED** [review]
Ok, I did a full pass here and I think it's close to landing. Nice work! Remaining comments are minor, I think, unless I've misunderstood something.

File: src/library_async.js
Comment (sbc100) [reply src/library_async.js:211]: Hmm.. I'm not really sure.  Perhaps you we want to use the `resolveGlobalSymbol` helper from library_dylink.js... although I'm sure what happens in the face of minification of imports/exports.. I guess that can't/doesn't happen with dynamic linking.

Comment (kripken) [reply src/library_async.js:211]: In that case, I think this PR is good to land? If we think of a better solution here we can do that later, but for now this seems good enough. And I don't have any other comments, unless you did @sbc100 ?

**Review by kripken — APPROVED** [review]

**Comment by kripken** [comment]
The test failures here should be looked at, however.

File: src/library_async.js
Comment (kripken) [inline src/library_async.js:281]: Please move the deps before the function itself (that is how we do it by convention).

File: tests/test_core.py
Comment (kripken) [reply tests/test_core.py:8203]: This still needs to be addressed.

**Review by kripken — COMMENTED** [review]
Looks good to land aside from 2 last things.

**Comment by kripken** [comment]
(The code size error on CI can be ignored - it is fixed on `main`. Please merge `main` while doing other changes here.)

File: tests/test_core.py
Comment (nokotan) [reply tests/test_core.py:8205]: That's correct.
I've added [some explanation](https://github.com/emscripten-core/emscripten/pull/15893/files#diff-841425e3398e112fdcaea671f046d9febda581c2ef8eb42e0cb28930df3aaaed)

File: site/source/docs/porting/asyncify.rst
Comment (kripken) [inline site/source/docs/porting/asyncify.rst:228]:
```suggestion
In the side module, you can compile sleep.cpp in the ordinary emscripten dynamic
```

File: site/source/docs/porting/asyncify.rst
Comment (kripken) [inline site/source/docs/porting/asyncify.rst:217]:
```suggestion
from other linked modules (and that will be on the stack in an async operation) should be listed in ``ASYNCIFY_IMPORTS``.
```

File: tests/test_core.py
Comment (kripken) [inline tests/test_core.py:3976]:
```suggestion
```
This is off by default, so this line is not needed.

**Review by kripken — COMMENTED** [review]
Lgtm with a few final comments. Also please merge in latest `main`.

**Comment by nokotan** [comment]
All suggested changes have been done, is there any more required modification before this PR merged?

File: tests/test_core.py
Comment (sbc100) [inline tests/test_core.py:3982]: Can you put open braces on the same line as the declaration?

Also this whole function look like it should be de-indented by 2 spaces.

File: tests/test_core.py
Comment (sbc100) [inline tests/test_core.py:3998]: Ditto

File: tests/test_core.py
Comment (sbc100) [inline tests/test_core.py:8169]: `self.set_setting('ASYNCIFY_IMPORTS', ['my_sleep'])`

File: tests/test_core.py
Comment (sbc100) [inline tests/test_core.py:3975]: Is this needed?

File: tests/test_core.py
Comment (sbc100) [inline tests/test_core.py:3993]: This files doesn't look like it uses iostream.  Should this be `stdio.h`?

**Review by sbc100 — APPROVED** [review]
lgtm % a few test nits

File: tests/test_core.py
Comment (sbc100) [reply tests/test_core.py:3975]: Sorry, I meant is `EXIT_RUNTIME` needed?

**Comment by zeuson0** [comment]
Hi, I want to know when will this PR be merged into the master, I need this for my project.

File: tests/test_core.py
Comment (sbc100) [reply tests/test_core.py:3982]: i.e. Can you remove `EXIT_RUNTIME` here and still have the test pass?

Comment (sbc100) [reply tests/test_core.py:3982]: If not perhaps add a comment about why its needed?

**Comment by kripken** [comment]
Merge conflict looks trivial so I resolved it.

**Comment by kripken** [comment]
PR lgtm assuming tests pass. @sbc100 you had a list question about whether `EXIT_RUNTIME` is needed in the test, but I think that's the last thing?

File: tests/test_core.py
Comment (nokotan) [reply tests/test_core.py]: It can be removed for now as the test case passes without `EXIT_RUNTIME`. (The test case had failed without `EXIT_RUNTIME` on the commit <https://github.com/emscripten-core/emscripten/pull/15893/commits/92148dcd3b09a96362555f25d6981a6c4e7262ba>, as far as I can remember.)

**Comment by kripken** [comment]
Ok, great, I think that resolves the last remaining issue here! Landing, thanks @nokotan

**Pull request merged**
